engine/crates/engine-v2/src/executor/coordinator.rs

```diff
@@ -0,0 +1,93 @@
+            engine,
+            operation_type: request_plan.operation_type,
+            operation_selection_set: request_plan.operation_selection_set,
+            response: Arc::new(Mutex::new(Response::new(
+                engine.schema.clone(),
+                request_plan.response_fields,
+            ))),
+            plans: request_plan.execution_plans,
+        }
+    }
+
+    pub async fn execute(&mut self) -> serde_json::Value {
+        let (sender, mut receiver) = futures_channel::mpsc::unbounded::<CoordinationTask>();
+        for plan_id in self.plans.all_without_dependencies() {
+            sender.unbounded_send(CoordinationTask::PlanStart(plan_id)).unwrap();
+        }
+        while let Some(task) = receiver.next().await {
+            match task {
+                CoordinationTask::PlanStart(plan_id) => {
+                    let plan = &self.plans[plan_id];
+                    if let Some(input) = self.response.lock().await.view(&plan.path, &plan.input) {
+                        let resolver = &self.engine.schema[plan.resolver_id];
+                        let execution_plan = Executor::build(
+                            self.engine,
+                            resolver,
+                            ExecutorRequest {
+                                operation_type: self.operation_type,
+                                response_objects: input,
+                                output: &plan.output,
+                            },
+                        );
+                        let sender = sender.clone();
+                        let proxy = ResponseProxy {
+                            inner: self.response.clone(),
+                        };
+                        async_runtime::spawn(async move {
+                            execution_plan.execute(proxy).await;
+                            sender.unbounded_send(CoordinationTask::PlanEnd(plan_id)).unwrap();
+                        });
+                    } else {
+                        sender.unbounded_send(CoordinationTask::PlanEnd(plan_id)).unwrap();
+                    }
+                }
+                CoordinationTask::PlanEnd(plan_id) => {
+                    // start the execution of any children for which all parents have finised.
+                    for executable_plan_id in self.plans.finished(plan_id) {
+                        sender
+                            .unbounded_send(CoordinationTask::PlanStart(executable_plan_id))
+                            .unwrap();
+                    }
+                    if self.plans.are_all_executed() {
+                        // We should data back through a stream to later support additional data
+                        // being sent like @defer or streaming
+                        break;
+                    }
+                }
+            }
+        }
+
+        // Response will be sent through a byte stream back. Client shouldn't be aware of wether we
+        // defer / stream / etc.
```

Comment:
> Client shouldn't be aware of whether we // defer / stream / etc.

Just a note for the future: I don't think this layer should deal with the byte-stream aspect of it.  This should probably return a stream of some `Response: Serialize` type, and leave the converting-to-byte-stream bit up to a higher layer.

Reply:
indeed, would definitely be better

Reply:
There's far too many different transports (and http libraries) for the core engine to be concerning itself with all those details